rootston/cursor.c

```diff
@@ -244,38 +244,33 @@ static void roots_cursor_press_button(struct roots_cursor *cursor,
 			roots_seat_begin_rotate(seat, view);
 			break;
 		}
-		return;
-	}
+	} else {
 
-	if (view && !surface) {
-		if (cursor->pointer_view) {
-			seat_view_deco_button(cursor->pointer_view, sx, sy, button, state);
+		if (view && !surface) {
+			if (cursor->pointer_view) {
+				seat_view_deco_button(cursor->pointer_view, sx, sy, button, state);
+			}
 		}
-	}
 
-	if (state == WLR_BUTTON_RELEASED &&
-			cursor->mode != ROOTS_CURSOR_PASSTHROUGH) {
-		cursor->mode = ROOTS_CURSOR_PASSTHROUGH;
-		if (seat->seat->pointer_state.button_count == 0) {
-			return;
+		if (state == WLR_BUTTON_RELEASED &&
+				cursor->mode != ROOTS_CURSOR_PASSTHROUGH) {
+			cursor->mode = ROOTS_CURSOR_PASSTHROUGH;
 		}
-	}
 
-	if (view && surface) {
-		if (!is_touch) {
-			wlr_seat_pointer_notify_button(seat->seat, time, button, state);
+		switch (state) {
+		case WLR_BUTTON_RELEASED:
+			if (!is_touch) {
+				roots_cursor_update_position(cursor, time);
+			}
+			break;
+		case WLR_BUTTON_PRESSED:
+			roots_seat_set_focus(seat, view);
+			break;
 		}
 	}
 
-	switch (state) {
-	case WLR_BUTTON_RELEASED:
-		if (!is_touch) {
-			roots_cursor_update_position(cursor, time);
-		}
-		break;
-	case WLR_BUTTON_PRESSED:
-		roots_seat_set_focus(seat, view);
-		break;
+	if (!is_touch) {
```

Comment:
Now when meta is pressed, this will fall through and send the button to the window which is wrong. Our implicit grab should eat the event.

Reply:
The problem was that wlr_seat_pointer_notify_button keeps track of how many buttons are down, and uses that to update the serial. If we don't want to send that event to the client, we need to split wlr_seat_pointer_notify_button in two different functions.

Reply:
I took a closer look at this, it seems that no pointer events are sent to the window, since the roots_seat_begin_* functions unset the pointer focus. Verified this with weston-clickdot.

Reply:
Can you give me steps to reproduce for the problem that this solves?

Reply:
Issue compared to master: press a mouse button on a window and release it outside. Now windows can no longer initiate a move or a resize.
If the previous issue is fixed, by doing everything this change does, but not call wlr_seat_pointer_notify_button when meta is pressed, then transforming a window with the compositor bindings does not allow a window to start a transformation.

Reply:
Can reproduce. Thanks for finding this.